### PR TITLE
04 Iteration

### PR DESCRIPTION
**Pull request by zentox**
Small enhancements:
- Separation of production code and test
- Profile pictures can be saved
- Course previews have been adjusted
- Course layout was adjusted
- Filtering of content types in a topic
- Versioning
   - Tracking content and course data
   - Added Content / Course change log option for detailed comments
   - Added Compare View
   - Layout of Compare History in course and content
   - History Previous, Next and Back to History buttons added
- The number of ratings under a content is displayed next to the rating.
- It is now possible to export PDF's

### CODE REVIEW

File: .github/workflows/pylint.yml [hunk 1]
```diff
@@ -39,6 +39,6 @@ jobs:
       - name: Run PyLint - frontend
         run: |
           pylint --fail-under=8 --load-plugins=pylint_django --django-settings-module=collab_coursebook.settings frontend
-      - name: Run PyLint - utils
+      - name: Run PyLint - test
         run: |
-          pylint --fail-under=8 --load-plugins=pylint_django --django-settings-module=collab_coursebook.settings utils
+          pylint --fail-under=8 --load-plugins=pylint_django --django-settings-module=collab_coursebook.settings test
```
Comment (LastNonak) [inline .github/workflows/pylint.yml:44]: CHANGED
- test app shall be tested via pylint instead of the utils app

File: base/views.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+"""Purpose of this file
+
+This file describes the frontend views.
```
Comment (LastNonak) [inline base/views.py:3]: Why is this file necessary? As far as i know, all frontend views should be located in the app corresponding to the frontend
-> maybe change the documentation of this file in order to make the purpose clear?

File: content/admin.py [hunk 3]
```diff
@@ -37,8 +38,12 @@ class ImageContentAdmin(CompareVersionAdmin):
 class LatexAdmin(CompareVersionAdmin):
     """LaTeX admin
 
-    Represents the LaTeX model in the admin panel.
+    Represents the Latex model in the admin panel.
+
+    :attr LatexAdmin.fields: Including fields into the form
```
Comment (LastNonak) [inline content/admin.py:43]: "Including fields into the form"? What does that mean?
-> do you mean "included form fields" or something like that?

File: base/views.py (hunk 2 above)
Comment (zentox) [reply base/views.py:3]: This file is from the original project and I only added a documentation of the file, so that really (almost) every file in the project is documented. So it is easier to find out what a file is used for or what a function etc. does. See for example in the module `export` the file `views.py`, which has relevant operations related to frontend view.

File: content/views.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+"""Purpose of this file
+
+This file describes the frontend views.
```
Comment (LastNonak) [inline content/views.py:3]: Same comment as above.
All frontend views have already been outsourced to the frontend app

File: export/templatetags/cc_export_tags.py [hunk 5]
```diff
@@ -19,34 +21,40 @@
...
+        return path + f"/{content_type}.tex"
     return path + "/invalid.tex"
 
 
 @register.filter
 def ret_path(value):
     """Return path to image
 
-        returns the correct path to the image in media directory
+    Returns the correct path to the image in media directory.
```
Comment (LastNonak) [inline export/templatetags/cc_export_tags.py:48]: maybe add that this returns only the absolute path

File: export/templatetags/cc_export_tags.py [hunk 6]
```diff
@@ -19,34 +21,40 @@
...
 
-        Parameters:
-            value
-        """
+    :param value: The path
+    :type value: str
 
-    # Compute the path to the image and escape \ for Latex
+    :return: the correct path to the image in media directory
```
Comment (LastNonak) [inline export/templatetags/cc_export_tags.py:53]: Same as above

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 7]
```diff
@@ -17,16 +17,50 @@ msgstr ""
...
+msgid "Sort by"
+msgstr "Sortieren nach"
+
 #: .\templates\frontend\base.html:28
-#, fuzzy
-#| msgid "Are you sure you want to delete the following comment?"
 msgid ""
 "Are you sure you want to change the language now? This will clear the form!"
-msgstr "Soll der folgende Kommentar wirklich gelöscht werden?"
+msgstr "Sicher die Sprache jetzt zu verändern? Die Form wird dann geleert!"
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:63]: Bessere Formulierung:
"Bist du sicher, dass du die Sprache ändern möchtest? Die Seite wird dann neu geladen"

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 8]
```diff
@@ -343,6 +432,98 @@ msgstr "Alle Kurse für den Zeitraum"
...
+"        "
+msgstr ""
+"\n"
+"            Vergleich <strong>%(date1)s</strong> mit <strong>%(date2)s</"
+"strong>:\n"
+"        "
+
+#: .\templates\frontend\history\history_compare_view.html:34
+msgid "There are no differences."
+msgstr "Es besteht keinen Unterschied."
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:482]: besser: "Es bestehen keine Unterschiede"

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 9]
```diff
@@ -343,6 +432,98 @@ msgstr "Alle Kurse für den Zeitraum"
...
+
+#: .\templates\frontend\history\history_compare_view.html:42
+msgid "Edit comment:"
+msgstr "Änderungsprotokoll: "
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:486]: Passt das so? Die englische und deutsche Formulierung sind schon stark unterschiedlich

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 10]
```diff
@@ -343,6 +432,98 @@ msgstr "Alle Kurse für den Zeitraum"
...
+"\n"
+"                Fields/entries marked with <sup class=\"follow\">*</sup> are "
+"not under reversion control.\n"
+"                It may be that not all marked information are correct.\n"
+"            "
+msgstr ""
+"\n"
+"                Mit <sup class=\"follow\">*</sup> gekennzeichneten Felder/"
+"Einträge sind nicht unter Kontrolle von Reversion. \n"
+"                Es kann sein, dass nicht alle gekennzeichnete Informationen "
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:507]: *gekennzeichneten

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 11]
```diff
@@ -343,6 +432,98 @@ msgstr "Alle Kurse für den Zeitraum"
...
+#: .\templates\frontend\history\history_compare_view.html:54
+msgid ""
+"\n"
+"                Fields/entries marked with <sup class=\"follow\">*</sup> are "
+"not under reversion control.\n"
+"                It may be that not all marked information are correct.\n"
+"            "
+msgstr ""
+"\n"
+"                Mit <sup class=\"follow\">*</sup> gekennzeichneten Felder/"
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:505]: *gekennzeichnete

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 12]
```diff
@@ -343,6 +432,98 @@ msgstr "Alle Kurse für den Zeitraum"
...
+"korrekt sind.\n"
+"            "
+
+#: .\templates\frontend\history\history_compare_view.html:64
+msgid "previous"
+msgstr "vorherige"
+
+#: .\templates\frontend\history\history_compare_view.html:70
+msgid "next"
+msgstr "nächst"
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:517]: *nächste (damit konsistent zu "vorherige")

File: frontend/locale/de_DE/LC_MESSAGES/django.po [hunk 13]
```diff
@@ -343,6 +432,98 @@ msgstr "Alle Kurse für den Zeitraum"
...
+
+#: .\templates\frontend\history\history_compare_view.html:75
+msgid "Revert to this version"
+msgstr "Auf diese Version zurückkehren"
+
+#: .\templates\frontend\history\history_compare_view.html:81
+msgid "Go back to history list"
+msgstr "Zurück zur Versionslist"
```
Comment (LastNonak) [inline frontend/locale/de_DE/LC_MESSAGES/django.po:525]: *Versionsliste

File: frontend/templates/frontend/course/view.html [hunk 14]
```diff
@@ -59,12 +61,18 @@ <h5 style="margin-top: 20px;">
...
+    {{ course.description | linebreaks }}
+
+
 
     <!-- Table of contents -->
     <div class="row" style="margin-top: 20px;">
         <div class="col-md-4">
             <ol class="list-group" style="font-weight: bold;">
-                <!-- ToDo Show nested structure -->
+                <!-- ToDo Click on links to jump to links -->
```
Comment (LastNonak) [inline frontend/templates/frontend/course/view.html:75]: Wurde in It. 5 dann erledigt

File: frontend/templates/frontend/profile/profile_edit.html [hunk 15]
```diff
@@ -8,7 +8,7 @@
 {% block content %}
     <h1>{% trans "Edit your profile" %}</h1>
 
-    <form method="POST" class="post-form" class="post-form">{% csrf_token %}
+    <form method="POST" class="post-form" enctype=multipart/form-data>{% csrf_token %}
```
Comment (LastNonak) [inline frontend/templates/frontend/profile/profile_edit.html:11]: csrf_token vllt. in nächste Zeile, um konsistent gegenüber anderen forms wie z.B. in addcontent.html zu bleiben
